paging/src/commonMain/kotlin/org/mobilenativefoundation/store/paging5/RealPager.kt

```diff
@@ -0,0 +1,158 @@
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.plus
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import org.mobilenativefoundation.store.core5.ExperimentalStoreApi
+import org.mobilenativefoundation.store.core5.StoreData
+import org.mobilenativefoundation.store.core5.StoreKey
+import org.mobilenativefoundation.store.store5.StoreReadResponse
+
+/**
+ * An internal class that implements the [Pager] interface.
+ * It manages the paging of data items based on the given [StoreKey].
+ * It also synchronizes updates to single items and collections.
+ *
+ * @param Id The type of the identifier that uniquely identifies data items.
+ * @param SK The subtype of [StoreKey.Single] that represents keys for single items.
+ * @param K The type of [StoreKey] used by the Store this pager delegates to.
+ * @param SO The subtype of [StoreData.Single] representing the data model for single items.
+ * @param O The type of [StoreData] representing the output of the Store this pager delegates to.
+ * @param scope The [CoroutineScope] within which the pager operates. Used to launch coroutines for data streaming and joining.
+ * @param streamer A [Streamer] function type that provides a flow of [StoreReadResponse] for a given key.
+ * @param joiner A [Joiner] function type that combines multiple paging data into a single [StateFlow].
+ * @param keyFactory A [KeyFactory] to create new [StoreKey] instances for single items based on their identifiers.
+ */
+@ExperimentalStoreApi
+internal class RealPager<Id : Any, SK : StoreKey.Single<Id>, K : StoreKey<Id>, SO : StoreData.Single<Id>, O : StoreData<Id>>(
+    private val scope: CoroutineScope,
+    private val streamer: Streamer<Id, K, O>,
+    private val joiner: Joiner<Id, K, SO>,
+    private val keyFactory: KeyFactory<Id, SK>
+) : Pager<Id, K, SO> {
+
+    // StateFlow to emit updates of PagingData.
+    private val mutableStateFlow = MutableStateFlow(emptyPagingData())
+    override val state: StateFlow<PagingData<Id, SO>> = mutableStateFlow.asStateFlow()
+
+    // Maps to keep track of all PagingData and corresponding streams.
+    private val allPagingData: MutableMap<K, PagingData<Id, SO>> = mutableMapOf()
+    private val allStreams: MutableMap<K, Job> = mutableMapOf()
+
+    // Mutexes for thread-safe access to maps.
+    private val mutexForAllPagingData = Mutex()
+    private val mutexForAllStreams = Mutex()
+
+    override fun load(key: K) {
+        if (key !is StoreKey.Collection<*>) {
+            throw IllegalArgumentException("Invalid key type.")
+        }
+
+        // Creating a child scope for coroutines.
+        val childScope = scope + Job()
+
+        // Launching a coroutine within the child scope for data loading.
+        childScope.launch {
+            // Locking the streams map to check and manage stream jobs.
+            mutexForAllStreams.withLock {
+                // Checking if there's no active stream for the key.
```

Comment:
just for readablility, it would be nicer to move these into helper methods.
especially, it is hard to visually track `mutexForAllStreams` usages without separate methods. 

```
launchChildCollectionLocked
```
kind of interface might be good.

alternatively, you can use a helper method to create the coroutine (lazily) and then only start it if you can get the lock (e.g. allStrams[key] is null).